source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2534 @@
+
+openUploadStream and openUploadStreamWithId
+```````````````````````````````````````````
+
+These operations SHOULD NOT be used in test files. See
+`IO operations for GridFS streams`_ in `Future Work`_. 
+
+
+.. _uploadFromStream:
+.. _uploadFromStreamWithId:
+
+uploadFromStream and uploadFromStreamWithId
+```````````````````````````````````````````
+
+The ``uploadFromStream`` and ``uploadFromStreamWithId`` operations' ``stream``
+parameter is a stream as defined in the `GridFS <../gridfs/gridfs-spec.rst>`__
+spec and not easily expressed in YAML/JSON. This parameter is expressed as an
+entity name, which the test runner MUST resolve to a `stream <entity_stream_>`_
+entity *before* passing it as a parameter to the method. The YAML file SHOULD
+use an `alias node`_ for a stream entity's ``id`` field
+(e.g. ``stream: *stream0``).
+
+
+changeStream
+~~~~~~~~~~~~
+
+Change stream entities are special in that they are not defined in
+`createEntities`_ but are instead created by using
+`operation.saveResultAsEntity <operation_saveResultAsEntity_>`_ with a
+`client_createChangeStream`_, `database_createChangeStream`_, or
+`collection_createChangeStream`_ operation.
+
+Test files SHOULD NOT use a ``watch`` operation to create a change stream, as
+the implementation of that method may vary among drivers. For example, some
+implementations of ``watch`` immediately execute ``aggregate`` and construct the
+server-side cursor, while others may defer ``aggregate`` until the change stream
+object is iterated.
+
+The `Change Streams <../change-streams/change-streams.rst>`__ spec does not
+define a consistent API for the ChangeStream class, since the mechanisms for
+iteration and capturing a resume token may differ between synchronous and
+asynchronous drivers. To account for this, this section explicitly defines the
+supported operations for change stream entities.
+
+
+iterateUntilDocumentOrError
+```````````````````````````
+
+Iterates the change stream until either a single document is returned or an
+error is raised.
+
+If `expectResult <operation_expectResult_>`_ is specified, it SHOULD be a single
+document.
+
+`Iterating the Change Stream <../change-streams/tests#iterating-the-change-stream>`__
+in the change stream spec cautions drivers that implement a blocking mode of
+iteration (e.g. asynchronous drivers) not to iterate the change stream
+unnecessarily, as doing so could cause the test runner to block indefinitely.
+This should not be a concern for ``iterateUntilDocumentOrError`` as iteration
+only continues until either a document or error is encountered.
+
+
+Special Test Operations
+-----------------------
+
+Certain operations do not correspond to API methods but instead represent
+special test operations (e.g. assertions). These operations are distinguished by
+`operation.object <operation_object_>`_ having a value of "testRunner". The
+`operation.name <operation_name_>`_ field will correspond to an operation
+defined below.
+
+
+failPoint
```

Comment:
I think we should note that drivers should set a reduced `heartbeatFrequencyMS`.

This came up recently in the context of CDRIVER-3793.

failPoints that mock a server [state change error](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#not-master-and-node-is-recovering) like `InterruptedAtShutdown` mark a server Unknown. And, normally, a 4.4+ server will immediately respond to any in-progress awaitable ismaster. But the failpoint itself does not actually simulate a state change. The server does not increment the topologyVersion (see SERVER-50500) or respond to in-progress awaitable ismaster commands. So it takes a full heartbeatFrequencyMS to rediscover the server. 

In chatting with @divjotarora and @mbroadst, it seems that drivers should reduce these values for retryable reads/writes tests so they do not take a full 10 seconds to run.

Note, there are some notes in other readmes:
- https://github.com/mongodb/specifications/blob/master/source/retryable-reads/tests/README.rst#speeding-up-tests which notes setting `minHeartbeatFrequencyMS` below the minimum
- https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/tests/README.rst#streaming-protocol-tests which notes reducing `heartbeatFrequencyMS` to 500ms for SDAM prose tests

I am not certain if setting a reduced `heartbeatFrequencyMS` for all tests in the unified test runner runs the risk of hiding bugs. If not, perhaps we should recommend drivers always set `heartbeatFrequencyMS` to 500ms? And if it is possible to reduce it below the minimum 500ms, they can reduce it further to speed up tests.

Reply:
I forget how other drivers do this (since PHP has no facility to do so via libmongoc), but is it typically a per-client change or a global change to the driver? To get to the point, I'm wondering if this is something we should discuss in the prose about implementing a test runner or allow test files to configure per-client in `createEntities`.

Note that there is no test file configuration for the internal MongoClient, which is just constructed from the URI (and whatever other options drivers want to inject). Since the internal MongoClient isn't used directly to test operations, I don't think there'd be a problem with it always using optimized values to minimize its recovery time.

Reply:
AFAIK, this is generally something that's done per-Client.

Reply:
> I forget how other drivers do this
If you're referring to overriding `heartbeatFrequencyMS` below 500ms, it is per-client in libmongoc, and I'd imagine the same for other drivers.

Reply:
Related: https://github.com/mongodb/specifications/pull/820

Reply:
@kevinAlbs @divjotarora and are we on the same page that this shouldn't affect the internal MongoClient (mainly used for `killAllSessions`, preparing collection data, and asserting outcome collections)? Fail points (both targeted and not) are set using client entities if that's relevant.

Reply:
I don't know enough about the unified format to say for sure, but it should only affect MongoClient instances that are used to set fail points. If there is a MongoClient that's only used for setup/teardown commands, it does not need this optimization.

Reply:
I do not think this should affect the internal MongoClient.

Reply:
Added language in "Executing a Test" to suggest lowering these values for clients that may set fail points, and noted that it must not override `heartbeatFrequencyMS` in a client's `uriOptions` (on the off chance a test files decides to specify that).

Please resolve if OK.